################################################################################
SHORT DESCRIPTION:
################################################################################
Move the extra QEMU arguments to the end of the list

################################################################################
LONG DESCRIPTION:
################################################################################
In general we think it makes more sense for "extra" arguments to be at the end.

Concretely, we want to be able to add drives using those extra arguments.
However, libxl assigns indexes to the drives, instead of letting qemu decide,
starting at index 0.
Without this patch, having a drive in extra would auto assign an index of 0 to
it, conflicting with the first drive xl adds...

################################################################################
CHANGELOG
################################################################################
Authors:
Chris Rogers <rogersc@ainfosec.com>

################################################################################
REMOVAL
################################################################################
Removing this will break atapi-pt.

################################################################################
UPSTREAM PLAN
################################################################################
We should probably upstream this.

################################################################################
INTERNAL DEPENDENCIES
################################################################################

################################################################################
PATCHES
################################################################################
Index: xen-4.10.0/tools/libxl/libxl_dm.c
===================================================================
--- xen-4.10.0.orig/tools/libxl/libxl_dm.c
+++ xen-4.10.0/tools/libxl/libxl_dm.c
@@ -636,8 +636,6 @@ static int libxl__build_device_model_arg
     if (state->saved_state) {
         flexarray_vappend(dm_args, "-loadvm", state->saved_state, NULL);
     }
-    for (i = 0; b_info->extra && b_info->extra[i] != NULL; i++)
-        flexarray_append(dm_args, b_info->extra[i]);
     flexarray_append(dm_args, "-machine");
     switch (b_info->type) {
     case LIBXL_DOMAIN_TYPE_PVH:
@@ -654,6 +652,8 @@ static int libxl__build_device_model_arg
     default:
         abort();
     }
+    for (i = 0; b_info->extra && b_info->extra[i] != NULL; i++)
+        flexarray_append(dm_args, b_info->extra[i]);
     flexarray_append(dm_args, NULL);
     *args = (char **) flexarray_contents(dm_args);
     flexarray_append(dm_envs, NULL);
@@ -1424,8 +1424,6 @@ static int libxl__build_device_model_arg
         flexarray_append(dm_args, "-incoming");
         flexarray_append(dm_args, GCSPRINTF("fd:%d",*dm_state_fd));
     }
-    for (i = 0; b_info->extra && b_info->extra[i] != NULL; i++)
-        flexarray_append(dm_args, b_info->extra[i]);
 
     flexarray_append(dm_args, "-machine");
     switch (b_info->type) {
@@ -1720,6 +1718,8 @@ end_search:
             flexarray_append(dm_args, user);
         }
     }
+    for (i = 0; b_info->extra && b_info->extra[i] != NULL; i++)
+        flexarray_append(dm_args, b_info->extra[i]);
     flexarray_append(dm_args, NULL);
     *args = (char **) flexarray_contents(dm_args);
     flexarray_append(dm_envs, NULL);
